src/sabre_swap/mod.rs

```diff
@@ -79,49 +87,331 @@ fn obtain_swaps(
+                }
+            }
+            if extended_set.len() >= EXTENDED_SET_SIZE {
+                done = true;
+                break;
+            }
+        }
+        tmp_front_layer = new_tmp_front_layer;
+    }
+    for node in decremented {
+        required_predecessors[node] += 1;
+    }
+    extended_set
+}
+
+fn cmap_from_neighor_table(neighbor_table: &NeighborTable) -> DiGraph<(), ()> {
+    DiGraph::<(), ()>::from_edges(neighbor_table.neighbors.iter().enumerate().flat_map(
+        |(u, targets)| {
+            targets
+                .iter()
+                .map(move |v| (NodeIndex::new(u), NodeIndex::new(*v)))
+        },
+    ))
+}
+
+/// Run sabre swap on a circuit
 ///
-/// Args:
-///     layers (EdgeList): The input layer edge list to score and find the
-///         best swaps
-///     layout (NLayout): The current layout
-///     neighbor_table (NeighborTable): The table of neighbors for each node
-///         in the coupling graph
-///     extended_set (EdgeList): The extended set
-///     distance_matrix (ndarray): The 2D array distance matrix for the coupling
-///         graph
-///     qubits_decay (QubitsDecay): The current qubit decay factors for
-///     heuristic (Heuristic): The chosen heuristic method to use
 /// Returns:
-///     ndarray: A 2d array of the best swap candidates all with the minimum score
+///     (SwapMap, gate_order): A tuple where the first element is a mapping of
+///     DAGCircuit node ids to a list of virtual qubit swaps that should be
+///     added before that operation. The second element is a numpy array of
+///     node ids that represents the traversal order used by sabre.
 #[pyfunction]
+pub fn build_swap_map(
+    py: Python,
+    num_qubits: usize,
+    dag: &SabreDAG,
+    neighbor_table: &NeighborTable,
+    distance_matrix: PyReadonlyArray2<f64>,
+    heuristic: &Heuristic,
+    seed: u64,
+    layout: &mut NLayout,
+    num_trials: usize,
+) -> (SwapMap, PyObject) {
+    let run_in_parallel = getenv_use_multiple_threads();
+    let dist = distance_matrix.as_array();
+    let coupling_graph: DiGraph<(), ()> = cmap_from_neighor_table(neighbor_table);
+    let outer_rng = Pcg64Mcg::seed_from_u64(seed);
+    let seed_vec: Vec<u64> = outer_rng
+        .sample_iter(&rand::distributions::Standard)
+        .take(num_trials)
+        .collect();
+    let (out_map, gate_order, best_layout) = if run_in_parallel {
+        (0..num_trials)
+            .into_par_iter()
+            .map(|trial_num| {
+                swap_map_trial(
+                    num_qubits,
+                    dag,
+                    neighbor_table,
+                    &dist,
+                    &coupling_graph,
+                    heuristic,
+                    seed_vec[trial_num],
+                    layout.clone(),
+                )
+            })
+            .min_by_key(|(out_map, _gate_order, _layout)| {
+                out_map.values().map(|x| x.len()).sum::<usize>()
+            })
```

Comment:
I'm debating changing this block to:

```suggestion
        let trial_results: Vec<(HashMap<usize, Vec<[usize; 2]>>, Vec<usize>, NLayout)> = (0..num_trials)
            .into_par_iter()
            .map(|trial_num| {
                swap_map_trial(
                    num_qubits,
                    dag,
                    neighbor_table,
                    &dist,
                    &coupling_graph,
                    heuristic,
                    seed_vec[trial_num],
                    layout.clone(),
                )
            }).collect();
            trial_results.into_iter().min_by_key(|(out_map, _gate_order, _layout)| {
                out_map.values().map(|x| x.len()).sum::<usize>()
            })
```
it is less efficient because it collects into an intermediate Vec. But I'm a bit concerned about potential non-determinism with the function because `min_by_key` will process the results when they finish meaning the output order may be dependent on execution speed.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/8572/commits/3d0429a974c3c91229974b8828c78c471d3e70cf